optimum/bettertransformer/models/bert/modeling_bert_fast.py

```diff
@@ -0,0 +1,122 @@
+        """
+        super().__init__()
+        # In_proj layer
+        self.in_proj_weight = nn.Parameter(
+            torch.cat(
+                [
+                    bert_layer.attention.self.query.weight,
+                    bert_layer.attention.self.key.weight,
+                    bert_layer.attention.self.value.weight,
+                ]
+            )
+        )
+        self.in_proj_bias = nn.Parameter(
+            torch.cat(
+                [
+                    bert_layer.attention.self.query.bias,
+                    bert_layer.attention.self.key.bias,
+                    bert_layer.attention.self.value.bias,
+                ]
+            )
+        )
+
+        # Out proj layer
+        self.out_proj_weight = bert_layer.attention.output.dense.weight
+        self.out_proj_bias = bert_layer.attention.output.dense.bias
+
+        # Linear layer 1
+        self.linear1_weight = bert_layer.intermediate.dense.weight
+        self.linear1_bias = bert_layer.intermediate.dense.bias
+
+        # Linear layer 2
+        self.linear2_weight = bert_layer.output.dense.weight
+        self.linear2_bias = bert_layer.output.dense.bias
+
+        # Layer norm 1
+        self.norm1_eps = bert_layer.attention.output.LayerNorm.eps
+        self.norm1_weight = bert_layer.attention.output.LayerNorm.weight
+        self.norm1_bias = bert_layer.attention.output.LayerNorm.bias
+
+        # Layer norm 2
+        self.norm2_weight = bert_layer.output.LayerNorm.weight
+        self.norm2_bias = bert_layer.output.LayerNorm.bias
+
+        # Model hyper parameters
+        self.num_heads = bert_layer.attention.self.num_attention_heads
+        self.embed_dim = bert_layer.attention.self.all_head_size
+
+        del bert_layer
+
+        # Last step: set the last layer to `False` -> this will be set to `True` when converting the model
+        self.is_last_layer = False
+
+    def forward(self, hidden_states, attention_mask, *_):
+        r"""
+        This is just a wrapper around the forward function proposed in:
+        https://github.com/huggingface/transformers/pull/19553
+        """
+        if hidden_states.is_nested:
+            attention_mask = None
+
+        if attention_mask is not None:
+            # attention mask comes in with values 0 and -inf. we convert to torch.nn.TransformerEncoder style bool mask
+            # 0->false->keep this token -inf->true->mask this token
+            attention_mask = attention_mask.bool()
+            attention_mask = torch.reshape(attention_mask, (attention_mask.shape[0], attention_mask.shape[-1]))
+            seqlen = attention_mask.shape[1]
+            lengths = torch.sum(~attention_mask, 1)
+            if not all([l == seqlen for l in lengths]):
+                hidden_states = torch._nested_tensor_from_mask(hidden_states, ~attention_mask)
+            attention_mask = None
+
+        hidden_states = torch._transformer_encoder_layer_fwd(
```

Comment:
It is possible but this will give us less control on the forward pass. See for example that for distilbert or bart, the forward pass behaves differently (regarding for example the variables names, or the usage of the attention mask). That is why I prefer to manually overwrite the forward pass for each model instead of calling the TransformerEncoderLayer forward directly for now

Reply:
Understood. Just want to make sure that there are no errors / unmeaningful results in case all the checks in the linked forward do not pass.

For examples, the checks
https://github.com/pytorch/pytorch/blob/0fc7de398636f4b53e6c3fde38b4e48a5ff5b37d/torch/nn/modules/transformer.py#L459
https://github.com/pytorch/pytorch/blob/0fc7de398636f4b53e6c3fde38b4e48a5ff5b37d/torch/nn/modules/transformer.py#L475
https://github.com/pytorch/pytorch/blob/0fc7de398636f4b53e6c3fde38b4e48a5ff5b37d/torch/nn/modules/transformer.py#L473 (`num_attention_heads` can be arbitrary)
https://github.com/pytorch/pytorch/blob/0fc7de398636f4b53e6c3fde38b4e48a5ff5b37d/torch/nn/modules/transformer.py#L467 (cost-free to check, maybe in `transform` (or submethod))

For the others I am not quite sure.

For the different variable names in bert / distilbert, this is code legacy from transformers? Basically we are bound to it?

Reply:
Ah yes, agreed that these tests are nice to have, let me think on how I can add them!
Regarding your second point, yes I think that we are bound to it

Reply:
I should have addressed them in 2fc26a58acc8a740eeead51301c380b826a4c46a ! Could you have a look 🙏

Reply:
all good